docs/_includes/Types.md

```diff
@@ -1511,21 +1511,131 @@ where _members_ can include fields, functions, methods and declarations of neste
 indeed refer to an object of allocated type `C`.
 The members of `J` are available as members
 of `C`.  A member in `J` is not allowed to be redeclared in `C`,
 except if the member is a non-`static` function or method without a
 body in `J`.  By doing so, type `C` can supply a stronger
-specification and a body for the member.
+specification and a body for the member. There is further discussion on
+this point in Section [Inheritance](#sec-inheritance).
 
 `new` is not allowed to be used with traits.  Therefore, there is no
 object whose allocated type is a trait.  But there can of course be
 objects of a class `C` that implement a trait `J`, and a reference to
 such a `C` object can be used as a value of type `J`.
 
+## Type `object`
+````grammar
+ObjectType_ = "object"
+````
+
+There is a built-in trait `object` that is a supertype of all
+reference types.
+Every class and every trait (other than `object` itself) automatically extends
+`object`. The purpose of type `object`
+is to enable a uniform treatment of _dynamic frames_. In particular, it
+is useful to keep a ghost field (typically named `Repr` for
+"representation") of type `set<object>`.
+
+It serves no purpose (but does no harm) to explicitly list `object` as
+an extendee in a class or trait declaration.
+
+Trait `object` contains no members.
+
+## Inheritance {#sec-inheritance}
+
+The purpose of traits is to be able to express abstraction: a trait
+encapsulates a set of behaviors; classes and traits that extend it
+_inherit_ those behaviors, perhaps specializing them.
+
+A trait or class may extend multiple other traits.
+The traits syntactically listed in a trait or class's `extends` clause
+are called its _direct parents_; the _transitive parents_ of a trait or class
+are its direct parents, the transitive parents of its direct parents, and
+the `object` trait (if it is not itself `object`).
+These are sets of traits, in that it does not matter if
+there are repetitions of a given trait in a class or trait's direct or
+transitive parents. However, if a trait with type parameters is repeated,
+it must have the same actual type parameters in each instance.
+Furthermore, a trait may not be in its own set of transitive parents; that is,
+the graph of traits connected by the directed _extends_ relationship may not
+have any cycles.
+
+A class or trait inherits (as if they are copied) all the instance members
+of its transitive parents. However, since names may not be overloaded in
+Dafny, different members (that is, members with different type signatures)
+within the set of transitive parents and the class or trait itself must have different names.[^overload]
+This restriction does mean that traits from different sources that
+coincidentally use the same name for different purposes cannot be combined
+by being part of the set of transitive parents for some new trait or class.
+
+Static members of a trait may not be redeclared and must be declared with a body.
```

Comment:
The body is only required if the program is compiled. That is, the type checker and verifier are fine if the static member has no body; it's only the compiler that would complain, as it would for any static or non-static member without a body. So, it is more accurate--and more to the point, here--to say that since static members of traits cannot be redeclared, any body of the member must be declared in the trait.

Reply:
Revised and clarified